setup/monitoring/stackdriver.md

```diff
@@ -4,9 +4,26 @@ title:  "Stackdriver"
 sidebar:
   nav: setup
 ---
-
 {% include toc %}
 
+### Regarding using Spinnaker Monitoring metrics directly as Stackdriver Monitoring Custom Metrics
+
+As of July 2019, it should be noted that multiple teams running Spinnaker on clusters in GKE have have opted to export metrics to DataDog, Prometheus or other tools rather than using the spinnaker-monitoring daemon to create StackDriver Custom Metrics from the spinnaker-monitoring daemon as described below.
+
+The reason for this is that the spinnaker-monitoring service was not initially designed to run on Stackdriver Monitoring and contains a large number of metrics. Google bills for each of these Custom Metrics individually, which can add *substantial* cost with the number of metrics present in the daemon. It's also problematic that installing the default dashboards in the daemon directly into a Stackdriver Monitoring Dashboard requires a manual step of communicating with Google to get access to the Stackdriver Dashboard API ([as described below](#installing-the-stackdriver-dashboards)).
```

Comment:
A small quibble here - it in fact _was_ designed with Stackdriver in mind (by @ewiseblatt), and he noted repeatedly (internally at least) that the cost to use it would be enormous (as you found out), and as such, added support for Prometheus and DataDog.

Reply:
I am not going to object to the wording here.

I am responding to background and to remind some courses of action about the issue.
To travis' point, I spent a few years raising these concerns and various other issues and constraints with them. The original implementation made some undesirable cost tradeoffs in order to work around other stackdriver issues and limitations at the time just to make it possible. Many of those have since been addressed.

Earlier this year I had made a proposal with the following reference implementation to the spinnaker community as a workaround for a few general issues, which was also motivated by mitigating stackdriver issues including costs. It would also make it feasible to create builtin metrics[*] in the future, which would remove all the custom metric costs, though getting approvals for builtin metrics has its own challenges at this time (maybe less so if there were community buy in)

https://github.com/ewiseblatt/spinnaker-monitoring/tree/experimental_transform/spinnaker-monitoring-daemon/experimental

I added the mechanisms that branch relies on into the spinnaker mainline several releases ago (1.11 or 1.12). That branch is just configuration that takes advantage of the mechanisms added to support it. There are no runtime code changes required. The enabling code changes are already in the product since Q1 if not since last year.

It greatly cuts down on the number of custom metric descriptors required and provides the means to control what you want on top of that. As you point out, custom metric descriptors are where the stackdriver costs come from. If I recall I had everything I wanted with just over 100 metric descriptors (Spinnaker internals and GCE, GKE, minimal AWS in a HA deployment) as compared to well over 600). Stackdriver gives you the first 500 free so depending on how much you are using elsewhere, that wouldnt add anything.

BTW, it might be the case that if you use prometheus with stackdriver then they do not charge for custom metrics. This is because the prometheus metrics are put into a different namespace so are technically not custom metrics (which impacts engineering tradeoffs because custom metrics are a more costly resource).

Having said that, if you are free to pick anything to monitor with, I like straight prometheus from among those choices. But if you do need to use stackdriver (e.g. you are monitoring other things with stackdriver and want a single monitoring system) then I think the proposal is the way to go. Personally would use the proposal regardless of metric store if you arent tied to the existing data model.

The branch could be moved into the mainline. The reasons it wasnt are not technical but this isnt the place to go into it. However if there was community buy in then I think it could be feasible to make the stackdriver costs go to 0 adding builtin metrics but there are a lot of policy hoops to jump through to get there.

[*] "builtin metrics" is not the current stackdriver term. I forget what they call them. By "builtin metrics" I mean the kind that are standard metric descriptor across all projects, not "custom metrics" which are specific metric descriptor to a particular project.

Reply:
Hi,

I noticed this discussion and need to comment as well... We are not using Stackdriver but Datadog - and the amount of custom metrics Spinnaker is sending is just enormous. (almost 30000 custom metrics a month), creating an extra invoice of $1000- about. Just wanted to say that Stackdriver's not the only one to "worry"...

I remember @ewiseblatt discussing this with you in Slack - and you pointed me to your experimental. I gave it a very quick go - but it broke basically all Spinnaker metrics in Datadog dashboards :) I did not have time to try to sort those out.

Going back to the roots of the issue - isn't there anything to be done - at "it's source", so the amount of custom metrics wouldn't be so explosively high - by default? It really is not a great feedback if someone just plugs in the monitoring - and you receive an interesting invoice next...